submitter/base.py

```diff
@@ -2,81 +2,136 @@
 from datetime import datetime
 from typing import Tuple, List
 
-from submission_broker.submission.submission import Submission, Entity
 
-from archiver import first_element_or_self
+from submission_broker.submission.submission import Entity
+from converter.biosamples import BioSamplesConverter
+from converter.biostudies import BioStudiesConverter
+from converter.ena.ena_study import EnaStudyConverter
+from converter.ena.ena_sample import EnaSampleConverter
+
+from archiver import first_element_or_self, ArchiveResponse, ConvertedEntity
+from hca.submission import HcaSubmission
 
 CREATED_ENTITY = 'CREATED'
 UPDATED_ENTITY = 'UPDATED'
 ERRORED_ENTITY = 'ERRORED'
 
 ARCHIVE_TO_HCA_ENTITY_MAP = {
-    'BioSamples': 'biomaterials',
-    'BioStudies': 'projects',
-    'ENA': 'projects'
+    'BioSamples': ['biomaterials'],
+    'BioStudies': ['projects'],
+    'ENA': ['projects', 'biomaterials']
+}
+
+CONVERTERS_BY_ARCHIVE_TYPES = {
+    'BioSamples_biomaterials': BioSamplesConverter(),
+    'BioStudies_projects': BioStudiesConverter(),
+    'ENA_projects': EnaStudyConverter(),
+    'ENA_biomaterials': EnaSampleConverter()
 }
 
 
 class Submitter(metaclass=ABCMeta):
-    def __init__(self, archive_client, converter):
+    def __init__(self, archive_client, converter, updater):
         self.archive_client = archive_client
         self.converter = converter
         self.release_date = None
+        self.updater = updater
 
-    def send_all_entities(self, submission: Submission, archive_type: str, error_key: str,
-                          additional_attributes: dict = None) -> Tuple[dict, List[str]]:
+    def convert_all_entities(self, submission: HcaSubmission, archive_type: str, additional_attributes: dict = None)\
+            -> List[ConvertedEntity]:
         if additional_attributes is None:
             additional_attributes = {}
-        response = {}
-        accessions = []
-        hca_entity_type = ARCHIVE_TO_HCA_ENTITY_MAP[archive_type]
-        for entity in submission.get_entities(hca_entity_type):
-            if archive_type == 'ENA' and hca_entity_type == 'projects':
-                self.__set_release_date_from_project(entity)
-            result, accession = self.send_entity(entity, archive_type, error_key, additional_attributes)
-            response.setdefault(result, []).append(entity)
-            accessions.append(accession)
-            additional_attributes.pop('accession', None)
-        return response, accessions
+        hca_entity_types = ARCHIVE_TO_HCA_ENTITY_MAP[archive_type]
+        converted_entities = []
+        for hca_entity_type in hca_entity_types:
+            is_update = False
+            if archive_type == 'ENA':
```

Comment:
When I see several occurrences of `if archive_type == 'ENA'` scattered in the function, it is a hint that the "strategy pattern" could (and probably should) have been used here.